src/test/java/bio/terra/service/dataset/DatasetBucketDaoTest.java

```diff
@@ -0,0 +1,246 @@
+import org.slf4j.LoggerFactory;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+@Category(Unit.class)
+public class DatasetBucketDaoTest {
+
+    private static final Logger logger = LoggerFactory.getLogger(DatasetBucketDaoTest.class);
+
+    @Autowired
+    private JsonLoader jsonLoader;
+
+    @Autowired
+    private DatasetDao datasetDao;
+
+    @Autowired
+    private DatasetBucketDao datasetBucketDao;
+
+    @Autowired
+    private ProfileDao profileDao;
+
+    @Autowired
+    private GoogleResourceDao resourceDao;
+
+    @Autowired
+    private NamedParameterJdbcTemplate jdbcTemplate;
+
+    @Autowired
+    private ResourceService resourceService;
+
+    private BillingProfileModel billingProfile;
+    private GoogleBucketResource bucketForFile;
+    private UUID projectId;
+    private Dataset dataset;
+
+    private UUID datasetId;
+    private UUID bucketResourceId;
+
+    private UUID createDataset(DatasetRequestModel datasetRequest, String newName) throws Exception {
+        datasetRequest.name(newName).defaultProfileId(billingProfile.getId());
+        dataset = DatasetUtils.convertRequestWithGeneratedNames(datasetRequest);
+        dataset.projectResourceId(projectId);
+        String createFlightId = UUID.randomUUID().toString();
+        UUID datasetId = UUID.randomUUID();
+        dataset.id(datasetId);
+        datasetDao.createAndLock(dataset, createFlightId);
+        datasetDao.unlockExclusive(dataset.getId(), createFlightId);
+        return datasetId;
+    }
+
+    private UUID createBucket() throws InterruptedException {
+        String ingestFileFlightId = UUID.randomUUID().toString();
+        bucketForFile =
+            resourceService.getOrCreateBucketForFile(
+                dataset.getName(),
+                billingProfile,
+                ingestFileFlightId);
+        return bucketForFile.getResourceId();
+    }
+
+    private UUID createDataset(String datasetFile) throws Exception  {
+        DatasetRequestModel datasetRequest = jsonLoader.loadObject(datasetFile, DatasetRequestModel.class);
+        return createDataset(datasetRequest, datasetRequest.getName() + UUID.randomUUID().toString());
+    }
+
+    @Before
+    public void setup() {
+        logger.info("-------------------Setup----------------------");
+        BillingProfileRequestModel profileRequest = ProfileFixtures.randomBillingProfileRequest();
+        billingProfile = profileDao.createBillingProfile(profileRequest, "hi@hi.hi");
+
+        GoogleProjectResource projectResource = ResourceFixtures.randomProjectResource(billingProfile);
+        projectId = resourceDao.createProject(projectResource);
+        logger.info("-------------------Test----------------------");
+    }
+
+    @After
+    public void teardown() {
+        logger.info("-------------------Cleanup----------------------");
+        datasetBucketDao.deleteDatasetBucketLink(datasetId, bucketResourceId);
+        datasetDao.delete(datasetId);
```

Comment:
I think we should put these in try blocks to ensure that the project gets deleted no matter what

Reply:
Yes, good idea!